### PR TITLE
FEATURE: API for sidebar

### PR DESCRIPTION
**Pull request by tgxworld**
API which allows extending Sidebar

### CODE REVIEW

**Comment by tgxworld** [comment]
I pushed my draft changes up for you to reference.

File: app/assets/javascripts/discourse/app/components/sidebar/section.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar/section.js:35]: Hmm can the `headerActions` be an object instead as I think it'll be a better data structure to use here as compared to the array since we need to lookup the header action by id.

File: app/assets/javascripts/discourse/app/lib/plugin-api.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:104]:
```suggestion
const PLUGIN_API_VERSION = "1.3.0";
```

No need to bump the version when we're adding an experimental API :)

File: app/assets/javascripts/discourse/app/lib/plugin-api.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:1687]: If we're adding a custom Sidebar section, `BaseSectionHeader` doesn't sound right. What do you think about `BaseCustomSidebarSection` instead?

File: app/assets/javascripts/discourse/app/lib/plugin-api.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:1744]: I think we should have an example where we add multiple links.

File: app/assets/javascripts/discourse/app/lib/sidebar/base-section-header.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-section-header.js:4]:
```suggestion
export default class BaseCustomSidebarSection {
```

File: app/assets/javascripts/discourse/app/lib/sidebar/base-section-link.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-section-link.js:4]:
```suggestion
export default class BaseCustomSidebarSectionLink {
```

File: app/assets/javascripts/discourse/app/lib/plugin-api.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/plugin-api.js:1750]:
```suggestion
  addSidebarSection(func) {
    addSidebarSection(func);
```

File: app/assets/javascripts/discourse/app/templates/components/sidebar/categories-section.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/categories-section.hbs:6]: I think we should be able to pass a non-array here and just let the API convert it into an array.

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section.hbs:48]:
```suggestion

    {{#if this.isMultipleHeaderActions}}
```

File: app/assets/javascripts/discourse/tests/acceptance/sidebar-test.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/tests/acceptance/sidebar-test.js:72]: This file is a getting a little crowded. I'll suggest splitting this into `sidebar-plugin-api-test` or something.

File: app/assets/javascripts/discourse/app/components/sidebar/section.js
Comment (lis2) [reply app/assets/javascripts/discourse/app/components/sidebar/section.js:35]: Do you mean for single action? Array is needed when multiple actions are provided, each button on the dropdown list can have a different action. But of course I can change that for single action we will have object.
```
    {{#if this.isMultipleHeaderActions}}
      <DropdownSelectBox
        @options={{hash icon=@headerActionsIcon placementStrategy="absolute"}}
        @content={{@headerActions}}
        @onChange={{action "handleMultipleHeaderActions"}}
        @class="edit-channels-dropdown"
      />
    {{/if}}
```

Comment (tgxworld) [reply app/assets/javascripts/discourse/app/components/sidebar/section.js:35]: I have some ideas but let's keep your changes for now because I'm not 100% sure if my idea will work. We can continuee to iterate after merging :)

File: app/assets/javascripts/discourse/app/components/sidebar/section-link.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar/section-link.js:5]: We're using Glimmer components going forward so let us change this to a Glimmer component.

File: app/assets/javascripts/discourse/app/components/sidebar/section-link.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar/section-link.js:7]:
```suggestion
  prefixCSS(color) {
```
```suggestion
  prefixCss(color) {
```

File: app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section-link.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section-link.js:61]:
```suggestion
  get SuffixCSSClass() {}
```
```suggestion
  get SuffixCssClass() {}
```

File: app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js:4]: The constructor for this section should also be documented.

File: app/assets/stylesheets/common/base/sidebar.scss
Comment (tgxworld) [inline app/assets/stylesheets/common/base/sidebar.scss:145]:
```suggestion
      font-size: var(--font-down-1);
```

File: app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js:135]:
```suggestion
    const actions = queryAll(
```

`$` is a jQuery identifier so we shouldn't be using it here since it isn't jQuery related.

File: app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js:149]:
```suggestion
    const links = queryAll(
```

Ditto: https://github.com/discourse/discourse/pull/17296/files#r919564731

File: app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/tests/acceptance/sidebar-plugin-api-test.js:17]: Do we need to clean up this state which we add to the app after the test?

File: app/assets/javascripts/discourse/app/components/sidebar.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:40]: I'm curious if we need this when each customSection is a component and can specify their own `willDestroy` hook.

File: app/assets/javascripts/discourse/app/components/sidebar/section-link.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar/section-link.js:6]: I'm curious why we need to cache this .

File: app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js:36]:
```suggestion
   * @returns {Array} Actions for header options button
```

File: app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/lib/sidebar/base-custom-sidebar-section.js:46]:
```suggestion
   * @returns {BaseCustomSidebarSectionLink[]} Links for section
```

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:11]: I wonder if we need a `@prefixCSSClass` here. Will just providing `sidebar-section-link-prefix-@prefixType` be sufficient?

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:19]:
```suggestion
          {{d-icon @prefixValue}}
```

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:13]: I think we should add CSS classes for these too so that it can easily be targeted with CSS.

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:37]:
```suggestion
    {{#if @suffixValue}}
```

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:38]: Same concern as https://github.com/discourse/discourse/pull/17296/files#r920811904

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section.hbs
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/templates/components/sidebar/section.hbs:30]: We need to fix the CSS class here because it is no longer a link.

**Review by tgxworld — COMMENTED** [review]
One last round of feedback but I think we're ready to merge after this.

File: app/assets/javascripts/discourse/app/components/sidebar.js
Comment (lis2) [reply app/assets/javascripts/discourse/app/components/sidebar.js:40]: Previously we had will Destroy, but I noticed it is never triggered. It it because customSection is POJO inherited from `BaseCustomSidebarSection`. That's why I decided to follow your teardown pattern

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (lis2) [reply app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:11]: No, I am using it to mark chat avatar with green ring when user is active:

Code from chat plugin
```
            get prefixCSSClass() {
              const activeUsers = this.chatService.presenceChannel.users;
              const user = this.channel.chatable.users[0];
              if (
                !!activeUsers?.findBy("id", user?.id) ||
                !!activeUsers?.findBy("username", user?.username)
              ) {
                return "active";
              }
              return "";
            }
```

Comment (tgxworld) [reply app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:11]: ooo icic. This is fine then.

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (lis2) [reply app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:38]: In suffix similarly, I am using class for new messages indicator. Blue for just new messages and green for mentions.

File: app/assets/javascripts/discourse/app/components/sidebar.js
Comment (tgxworld) [reply app/assets/javascripts/discourse/app/components/sidebar.js:40]: O sorry, in this case we need to define a `willDestroy` hook in `<Sidebar::Section>` component and pass in the function that the `willDestroy` hook will call.

File: app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs
Comment (tgxworld) [reply app/assets/javascripts/discourse/app/templates/components/sidebar/section-link.hbs:38]: Got it :+1:

File: app/assets/javascripts/discourse/app/components/sidebar/section-link.js
Comment (lis2) [reply app/assets/javascripts/discourse/app/components/sidebar/section-link.js]: You are right, no need for that :)

**Review by tgxworld — COMMENTED** [review]
Looks good :+1: Thank you for the work here and let's ship this :)

**Comment by tgxworld** [comment]
@lis2 O I just realised you have to be the one approving it cause I created the PR xD

**Review by lis2 — APPROVED** [review]

**Pull request merged**
